crates/wasi-common/src/snapshots/preview_1.rs

```diff
@@ -895,41 +934,67 @@ impl wasi_snapshot_preview1::WasiSnapshotPreview1 for WasiCtx {
         let oflags = OFlags::from(&oflags);
         let fdflags = FdFlags::from(fdflags);
         let path = path.as_str()?;
-        if oflags.contains(OFlags::DIRECTORY) {
-            if oflags.contains(OFlags::CREATE)
+
+        let mut required_caps = DirCaps::OPEN;
+        if oflags.contains(OFlags::CREATE) {
+            required_caps = required_caps | DirCaps::CREATE_FILE;
+        }
+        let dir = dir_entry.get_cap(required_caps)?;
+
+        // `O_DIRECTORY` is incompatible with `O_CREAT`, `O_EXCL`, and `O_TRUNC`.
+        if oflags.contains(OFlags::DIRECTORY)
+            && (oflags.contains(OFlags::CREATE)
```

Comment:
(nit, I suppose?) I think this if statement is way too complex and hard to read, which is exactly why I introduced https://github.com/rvolosatovs/wasmtime/blob/5ffb7f8990d1562bf9a1b6b29e3faadc934d024f/crates/wasi-common/src/snapshots/preview_1.rs#L895-L898 in #4947